content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+of interfaces. Since the interfaces for an instruction are clear, basic blocks will be an extension.
+It can be seen as a part of a function for the purposes of granularity. So a (loosely coupled) accelerator
+can work at a higher granularity than an instruction.
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+      "bb_replace", 
+      Type::getInt32Ty(Ctx),
+      Type::getInt32Ty(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    /* Add runtime basic block selection here */
+    
+    /* End of selection algorithm */
+    
+    /* Select a basic block */
+    BasicBlock* workList = NULL; // collects replaced basic block within iterator
+    /* mark basic blocks for deletion */
+    for (auto& B : F) {
+        for (auto& I : B) {
+            workList = &B;
+        }
+    }
+    
+    /* collect information on the basic block */
+    BasicBlock* predBB = workList->getSinglePredecessor();
+    BasicBlock* succBB = workList->getSingleSuccessor();
+    
+    /* delete basic block being replaced */
+    if (workList != NULL) {
+        workList->eraseFromParent();
+    }
+    
+    /* create new basic block */ 
+    BasicBlock *pb = BasicBlock::Create(Ctx, "simple", &F);
+    
+    /* create the runtime function call */
+    IRBuilder<> builder(pb);
+    Constant *i32_configs = ConstantInt::get(Type::getInt32Ty(Ctx), /*configuration*/, true);
+    Value* args[] = {i32_configs};
+    Value* ret =  builder.CreateCall(rtFunc, args);
+    
+    /* create a jump to next block */
+    Instruction *brInst = BranchInst::Create(succBB, pb);
+    
+    /* update jump from previous block */
+    Instruction *brPred = predBB->getTerminator();
+    brPred->setSuccessor(0, pb);
+```
+However, this also conflates challenges from both prior levels of abstraction.
+Unlike a single instruction, it's highly program dependent how a basic block interacts with
+rest of the program. Unlike a function, it is not clear what those program specific interfaces
+are, and has to be created at an intermediate-representation level. To make matters worse,
+these interfaces are intermediate-repesentation specific. If you target a specific optimization
+(say -O1), your pass will probably need updating for a different optimization.
+This is all great, if you're focusing on the compiler optimizations. But for a hardware 
+accelerator simulation nothing can be worse, now you are entangled in the compiler optimization
+with the simulator for performance.
+
+## Replacing chains of instructions
+One way to strike the middle ground provided by basic blocks, without the baggage from a
+particular compiler optimization, is to critically select a set of instructions to 
+replace that gives you sufficient coarseness in granularity but also decouples from the 
+intermediate representation.
```

Comment:
As we discussed in person, it is not really right to say that these transformations are tied to specific optimization levels. They might be tied to code patterns, which may be affected by optimization levels. But it's important to be precise about what those patterns in the code are that you're talking about. Please be specific here.